lib/charms/lego_base_k8s/v1/lego_client.py

```diff
@@ -0,0 +1,431 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""# lego_client Library.
+
+This library helps user create charms that use LEGO, a letsencrypt client, to provide
+certificates to other charms in the charm ecosystem. The certificate requesting functionality
+and the distribution of issued certificates are all done by the Charm class that's provided
+in this metaclass, which users can inherit to get all of its functionality.
+
+Some configuration by the inheriting charm is required.
+
+## Getting Started
+To get started using the library, you need to fetch the library using `charmcraft`.
+```shell
+charmcraft fetch-lib charms.lego_client_operator.v1.lego_client
+```
+
+You will also need to fetch the following libraries:
+
+```shell
+charmcraft fetch-lib charms.tls_certificates_interface.v4.tls_certificates
+charmcraft fetch-lib charms.certificate_transfer_interface.v1.certificate_transfer
+charmcraft fetch-lib charms.loki_k8s.v1.loki_push_api
+```
+
+You will then need to add the following library to the charm's `requirements.txt` file
+alongside the default requirements needed by charms:
+- cryptography
+- cosl
+- pylego
+
+You will need to add the following to your charmcraft.yaml file:
+```yaml
+config:
+  options:
+    email:
+      type: string
+      description: Account email address to receive notifications from Let's Encrypt.
+    server:
+      type: string
+      description: Certificate authority server
+      default: "https://acme-v02.api.letsencrypt.org/directory"
+    <plugin-name>-config-secret-id:
+      type: string
+      description: The secret id of the secret that contains all of the configuration options
+        required to get a certificate. The secret should be in key-value form, with the keys
+        being the environment variables in lowercase kebab-case, and the values being the values
+        that are required for Lego.
+
+        For example, the required variables for Cloudflare is documented in https://go-acme.github.io/lego/dns/cloudflare/index.html
+        The format for the secret of a cloudflare plugin should be
+        ```
+        {
+            "cloudflare-email": "your@email.com",
+            "cloudflare-api-key": "yourapikey123"
+        }
+        ```
+        A user can create a secret in this format by running
+        ```
+        juju add-secret plugin-credentials cloudflare-email="your@email.com" cloudflare-api-key="yourapikey123"
+        ```
+
+provides:
+  certificates:
+    interface: tls-certificates
+  send-ca-cert:
+    interface: certificate_transfer
+requires:
+  logging:
+    interface: loki_push_api
+
+parts:
+  charm:
+    build-snaps:
+      - go
```

Comment:
Won't we need a bit more to build cryptography?

```
parts:
  charm:
    build-packages:
      - libffi-dev
      - libssl-dev
      - rustc
      - cargo
      - pkg-config
```

Reply:
yes, these will be needed for the tls-certificates-interface lib, just didn't know if we should repeat them here. Added